catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/osgi/CdmServiceWaiter.java

```diff
@@ -0,0 +1,175 @@
+
+  private final ScheduledExecutorService scheduledExecutorService;
+
+  private final TimeUnit timeUnit;
+
+  private final long timeout;
+
+  private int checkPropsPeriod = 30;
+
+  private TimeUnit checkPropsTimeUnit = TimeUnit.SECONDS;
+
+  private Set<String> propertyValues;
+
+  public CdmServiceWaiter(String propertyId, String className, CdmState cdmState) {
+    this(propertyId, className, cdmState, DEFAULT_TIMEOUT_MINUTES);
+  }
+
+  public CdmServiceWaiter(String propertyId, String className, CdmState cdmState, long timeout) {
+    this(propertyId, className, cdmState, timeout, TimeUnit.MINUTES);
+  }
+
+  public CdmServiceWaiter(
+      String propertyId, String className, CdmState cdmState, long timeout, TimeUnit timeUnit) {
+    this(
+        propertyId,
+        className,
+        timeout,
+        timeUnit,
+        Executors.newSingleThreadScheduledExecutor(
+            StandardThreadFactoryBuilder.newThreadFactory("CdmServiceWaiter")),
+        new OsgiServices(),
+        cdmState);
+  }
+
+  @VisibleForTesting
+  public CdmServiceWaiter(
+      String propertyId,
+      String className,
+      long timeout,
+      TimeUnit timeUnit,
+      ScheduledExecutorService scheduledExecutorService,
+      OsgiServices osgiServices,
+      CdmState cdmState) {
+    this.propertyId = propertyId;
+    this.timeout = timeout;
+    this.timeUnit = timeUnit;
+    this.className = className;
+    this.scheduledExecutorService = scheduledExecutorService;
+    this.osgiServices = osgiServices;
+    this.initState = cdmState;
+  }
+
+  /**
+   * This method will block until all property values have been satisfied by any OSGi services
+   * registered under className which contain the given service property key.
+   *
+   * @param propertyValues service property values to wait for
+   * @throws InterruptedException if the waiting thread was interrupted before finding all service
+   *     property values
+   */
+  public void waitFor(Set<String> propertyValues) throws InterruptedException {
+    if (propertyValues.isEmpty()) {
+      return;
+    }
+
+    synchronized (initState) {
+      this.propertyValues = new HashSet<>(propertyValues);
+      LOGGER.trace(
+          "Waiting up to {} {} for services with class name {}, containing property key {}, with property values {}",
+          timeout,
+          timeUnit,
+          className,
+          propertyId,
+          propertyValues);
+
+      try {
+        scheduledExecutorService.scheduleAtFixedRate(
+            this::checkPropertyAgainstValues, 0, checkPropsPeriod, checkPropsTimeUnit);
+        initState.wait(timeUnit.toMillis(timeout));
+      } finally {
+        scheduledExecutorService.shutdownNow();
+      }
+    }
+  }
+
+  private void checkPropertyAgainstValues() {
+    BundleContext bundleContext = osgiServices.bundleContextFor(this.getClass());
+    List<ServiceReference> serviceReferences =
+        osgiServices.getServiceReferences(bundleContext, className);
```

Comment:
❓ Couldn't you do this with an ldap filter instead of dealing with the property keys directly?